README.md
<img alt="Maikor" src="https://github.com/MaikorAppPublic/.github/raw/main/profile/controller_with_logo_blue.png?raw=true" width="200">

>⚠️ Work in progress
>
> Links may be broken, features missing, etc

*Cross-platform 16-bit game system*

See more at [maikor.app](https://maikor.app) and the [project homepage](https://github.com/MaikorAppPublic)

### Play

[Android](https://github.com/MaikorAppPublic/android-app)

[iOS](https://github.com/MaikorAppPublic/ios-app)

[Windows, macOS and Linux](https://github.com/MaikorAppPublic/desktop-app)

### Make

[iOS IDE](https://github.com/MaikorAppPublic/ios-app)

[Desktop IDE](https://github.com/MaikorAppPublic/desktop-ide)

[Build tools](https://github.com/MaikorAppPublic/build-tools)

[REPL](https://play.vm.maikor.app)

[Docs](https://docs.maikor.app)

## maikor-platform

- constants
  - Graphical constants and limits for the platform
- graphics
  - Sprite bit masks
- input
  - Bits used for inputs
- mem
  - sizes
    - Size in byte of sections
  - addresses
    - Address in bytes of sections
- names
  - op
    - Short names, eg. `CPY.B`
  - full
    - OP + args, eg. `CPY.B (R,N)`
- op_params
  - Registers are encoded in one byte, the first four bits are Pre/Post Inc/Dec and the last four bits are the register
- ops
  - Codes for ops
- registers
  - offset
    - Address of register
  - ids
    - ID of register
  - flags
    - Bits for flag register

CHANGELOG.md
## Pre-alpha

### 0.1.29
- Add line draw interrupt (this will fire 60x per frame, but interval isn't guaranteed)
- Fix bugs with interrupts

### 0.1.28
- *BREAKING CHANGE*
- Due to miscalculation controller graphics memory size was half of what was needed and there isn't enough memory left to fix this, so:
  - Controller graphics and controller type memory addresses removed
  - Controller type interrupt still works
  - Controller graphics palette increased to 8 colours
  - Controller graphics will be drawn by VM, controlled by controller graphics table
  - Memory addresses of later sections moved
- Removed line drawn interrupt as it won't be used

### 0.1.27
- Move interrupts addresses to start at x200 and reduce size to 32 bytes per interrupt

### 0.1.26
- *BREAKING CHANGE*
- Split code and ram banks into 2
- Add one extra controller sprite to table

### 0.1.25
- Actually add BMUL to ALL

### 0.1.24
- Fix bug with `is_special_memory`
- Add BMUL

### 0.1.23
- Add RCL and RCR

### 0.1.22
- *BREAKING CHANGE* (minor)
- Change input::mask::* types to u8 from usize
- Add wave table memory address and size

### 0.1.21
- Add MSWP for easy sprite management

### 0.1.20
- *BREAKING CHANGE*
- Atlas count changed from 2 to 4
- Atlas bank size changed from 8000 to 4000
- Bit mask and order changed for sprites, layer headers and contents
- Each sprite can use any atlas
- Add structs to make working with sprites and layers easier

### 0.1.19
- *BREAKING CHANGE*
- Increase stack to 1000b and change it's address
- Change total sizes to usize so they are the correct value of 65536 instead of 65535

### 0.1.18
- *BREAKING CHANGE*
- Fix issue with offset addressing (split reg into reg and ext reg)

### 0.1.17
- *BREAKING CHANGE* (MAJOR)
- Rename repo and dep to `platform`
- Remove useless op_byte_count method (offset addressing means that byte count is controlled by params now)
- Add datetime interrupt (it fires on date time changes greater than 1s, for example after resuming a mobile host or loading from a save state)

### 0.1.16
- Add PUSH and POP instructions at x70
- Add reg and num offset addressing

### 0.1.15
- Add count for controller types

### 0.1.14
- *BREAKING CHANGE*
- Add `DATETIME` and `RAND` mem addresses
- Allow 16 colours per palette by making 0,0,0 be transparent
- Most memory addresses increase 12 bytes
- Add bit masks for graphics

### 0.1.13
- *BREAKING CHANGE* (minor)
- Removed `SAVE_COUNT`, use `constants::SAVE_COUNT`
- Add method to check if mem address is expensive to update (for example if it controls a bank)
- Add `SAVE_CONTROL` 

### 0.1.12

- Add `EHALT` command
- Add `MAY_JMP_OPS` and `MUST_JMP_OPS` lists

### 0.1.11

- *BREAKING CHANGE*
- Add interrupt control
- Add controller sprite palette
- Add controller graphic sprite table + IDs
- Reduce background layer to 3 (from 4)
- Increase code and ram bank size to 8700b (from 8192)
- Increase stack to 900b (from 800)
- Increase reserved to 171b (from 37)
- Add interrupt control flags
- Add save count constant
- Increase sound to 30b (from 25)
- Add 4 keyboard controller types (for different layouts)

### 0.1.10

- Add SLEEP
- Add internal addresses (VLine)
- Add X,Y buttons
- Add controller type
- Add controller graphic addresses

### 0.1.9

- *BREAKING CHANGE*
- Fix mem TOTAL, by changing it to 0xFFFF 
- Add timer and interrupt addresses
- Reduce stack to 800 bytes

### 0.1.8

- *BREAKING CHANGE*
- Move register offset to VM
  - As language registers should be independent of VM implementation

### 0.1.7

- Correct typos in names

### 0.1.6

- Add register::names
- Make ppid fields public

### 0.1.5

- *BREAKING CHANGE*
- Change register method names to make purpose clearer
- Add registers::offset::from_id method
- Add struct for register apps, with to/from u8

### 0.1.4

- *BREAKING CHANGE*
- Replace panics with results

### 0.1.3

- *BREAKING CHANGE*
- Add more commands
  - ADDC
  - SUBC
- Remove half carry flag, left as reserved for now
- Mark 13 bytes from stack as reserved, stack is 850 bytes

### 0.1.2

- Add missing commands   
  - JBS
  - JBC
- Correct OP names

### 0.1.1

- *BREAKING CHANGE*
- Move code from unnecessary `mod`

### 0.1.0

- Initial release
- Moved from `maikor-vm` as ops, etc will be needed by projects that don't need VM structs, etc

src/mem.rs
pub mod sizes {
    use crate::constants::{LAYER_COUNT, PALETTE_COUNT, SPRITE_COUNT};
    use crate::input::controller_type;

    pub const CODE_BANK: u16 = 4200;
    pub const RAM_BANK: u16 = 4200;
    pub const MAIN_CODE: u16 = 9000;
    pub const MAIN_RAM: u16 = 9000;
    pub const SOUND: u16 = 30;
    pub const WAVE_TABLE: u16 = 16;
    //Byte 0 is direction, byte 1 is action
    pub const INPUT: u16 = 2;
    pub const CODE_BANK_ID: u16 = 1;
    pub const RAM_BANK_ID: u16 = 1;
    pub const SAVE_BANK_ID: u16 = 1;
    pub const SAVE_BANK: u16 = 4096;
    pub const SAVE_CONTROL: u16 = 1;
    pub const SPRITE: u16 = 5;
    //255 sprites, each taking 5 bytes
    //(8b X, 8b Y, 8b tile id, 1b flipV, 1b flipH, 2b palette, 1b src size, 2b order, 1b half alpha, 1b rotated, 1b enabled, 2b atlas, 3b ?)
    pub const SPRITE_TABLE: u16 = SPRITE_COUNT as u16 * SPRITE;
    //4 layers, each header is made of 3 bytes (8b X, 8b Y, 1b visible, 2b atlas)
    pub const LAYERS_HEADER: u16 = 3 * LAYER_COUNT as u16;
    //4 layers, each made of 1320 (44x30) tiles, each made of 2 bytes (8b tile id, 1b flipV, 1b flipH, 2b palette, 1b half alpha, 1b rotated, 2b ?)
    pub const LAYERS_CONTENT: u16 = 1320 * 2 * LAYER_COUNT as u16;
    //4 palettes, each made of 16 colors, each color is 3 bytes
    pub const PALETTE: u16 = 16 * 3; //0,0,0 is transparent
    pub const PALETTES_TOTAL: u16 = PALETTE * PALETTE_COUNT as u16;
    //25x10 (tiles) 200x160 (pixels) 100x160 (bytes) atlas of palette index (two colour IDs per byte)
    pub const ATLAS: u16 = 100 * 40;
    pub const ATLAS_BANK_ID: u16 = 1;
    pub const STACK: u16 = 1000;
    pub const SP: u16 = 2;
    pub const FP: u16 = 2;
    pub const TIMER_CONTROL: u16 = 2;
    pub const TIMER_VALUE: u16 = 1;
    pub const TIMERS: u16 = TIMER_CONTROL + TIMER_VALUE * 4;
    pub const VLINE: u16 = 1;
    pub const IRQ_RET_ADDR: u16 = 2;
    pub const IRQ_REG_DUMP: u16 = 8;
    pub const IRQ_CONTROL: u16 = 1;
    pub const IRQ_INTERNAL: u16 = IRQ_RET_ADDR + IRQ_REG_DUMP + IRQ_CONTROL;
    //year (0=2000), month (1 = jan), day, hour (24), min, sec
    pub const DATETIME: u16 = 6;
    pub const RAND: u16 = 1;
    //0,0,0 is transparent
    pub const CONTROLLER_PALETTE: u16 = 3 * 8;
    //8 sprites, each taking 3 bytes
    //(bits) 8 x, 8 y, 1 enabled, 4 id, 3 ?
    pub const CONTROLLER_TABLE: u16 = 3 * controller_type::COUNT as u16;
    pub const CONTROLLER_TOTAL: u16 = CONTROLLER_PALETTE + CONTROLLER_TABLE;
    pub const ATLAS_TOTAL: u16 = (ATLAS + ATLAS_BANK_ID) * 4;
    pub const LAYER_TOTAL: u16 = LAYERS_CONTENT + LAYERS_HEADER;
    pub const GRAPHICS_TOTAL: u16 =
        LAYER_TOTAL + SPRITE_TABLE + PALETTES_TOTAL + ATLAS_TOTAL + CONTROLLER_TOTAL;
    pub const SYSTEM_TOTAL: u16 = MAIN_CODE
        + MAIN_RAM
        + CODE_BANK
        + CODE_BANK
        + RAM_BANK
        + RAM_BANK
        + CODE_BANK_ID
        + CODE_BANK_ID
        + RAM_BANK_ID
        + RAM_BANK_ID
        + STACK
        + SP
        + FP
        + TIMERS
        + IRQ_INTERNAL
        + VLINE;
    pub const HARDWARE_TOTAL: u16 =
        WAVE_TABLE + SOUND + INPUT + SAVE_BANK_ID + SAVE_BANK + SAVE_CONTROL + DATETIME + RAND;
    pub const RESERVED: u16 = 106;
    pub const TOTAL: usize =
        (GRAPHICS_TOTAL + SYSTEM_TOTAL + HARDWARE_TOTAL) as usize + RESERVED as usize;
}

#[rustfmt::skip]
pub mod address {
    pub const CODE: u16 = 0x0; //0
    pub const CODE_BANK_1: u16 = 0x2328; //9000
    pub const CODE_BANK_2: u16 = 0x3390; //13200
    pub const RAM: u16 = 0x43F8; //17400
    pub const RAM_BANK_1: u16 = 0x6720; //26400
    pub const RAM_BANK_2: u16 = 0x7788; //30600
    pub const INPUT: u16 = 0x87F0; //34800
    pub const SOUND: u16 = 0x87F2; //34802
    pub const SAVE_BANK_ID: u16 = 0x8810; //34832
    pub const SAVE_BANK: u16 = 0x8811; //34833
    pub const ATLAS1: u16 = 0x9811; //38929
    pub const ATLAS2: u16 = 0xA7B1; //42929
    pub const ATLAS3: u16 = 0xB751; //46929
    pub const ATLAS4: u16 = 0xC6F1; //50929
    pub const PALETTES: u16 = 0xD691; //54929
    pub const SPRITE_TABLE: u16 = 0xD751; //55121
    pub const LAYER_HEADERS: u16 = 0xDC4C; //56396
    pub const LAYERS: u16 = 0xDC55; //56405
    pub const CODE_BANK_1_ID: u16 = 0xFB45; //64325
    pub const RAM_BANK_1_ID: u16 = 0xFB46; //64326
    pub const ATLAS1_BANK_ID: u16 = 0xFB47; //64327
    pub const ATLAS2_BANK_ID: u16 = 0xFB48; //64328
    pub const ATLAS3_BANK_ID: u16 = 0xFB49; //64329
    pub const ATLAS4_BANK_ID: u16 = 0xFB4A; //64330
    pub const SP: u16 = 0xFB4B; //64331
    pub const FP: u16 = 0xFB4D; //64333
    pub const TIMER_CONTROL: u16 = 0xFB4F; //64335
    pub const TIMER_VALUE1: u16 = 0xFB51; //64337
    pub const TIMER_VALUE2: u16 = 0xFB52; //64338
    pub const TIMER_VALUE3: u16 = 0xFB53; //64339
    pub const TIMER_VALUE4: u16 = 0xFB54; //64340
    pub const IRQ_RET_ADDR: u16 = 0xFB55; //64341
    pub const IRQ_REG_DUMP: u16 = 0xFB57; //64343
    pub const CONTROLLER_PALETTE: u16 = 0xFB5F; //64351
    pub const CONTROLLER_TABLE: u16 = 0xFB77; //64375
    pub const IRQ_CONTROL: u16 = 0xFB92; //64402
    pub const SAVE_CONTROL: u16 = 0xFB93; //64403
    pub const DATETIME: u16 = 0xFB94; //64404
    pub const RAND: u16 = 0xFB9A; //64410
    pub const WAVE_TABLE: u16 = 0xFB9B; //64411
    pub const CODE_BANK_2_ID: u16 = 0xFBAB; //64427
    pub const RAM_BANK_2_ID: u16 = 0xFBAC; //64428
    pub const VLINE: u16 = 0xFBAD; //64429
    pub const RESERVED: u16 = 0xFBAE; //64430
    pub const STACK: u16 = 0xFC18; //64536
    pub const MAX: u16 = 0xFFFF; //65535

    pub mod interrupt {
        pub const IRQ_INPUT: u16 = 0x0200; //512
        pub const IRQ_SCREEN_DRAW: u16 = 0x0220; //544
        pub const IRQ_TIMER: u16 = 0x0240; //576
        pub const IRQ_CONTROLLER: u16 = 0x0260; //608
        pub const IRQ_DATETIME: u16 = 0x0280; //640
        pub const IRQ_LINE_DRAW: u16 = 0x02A0; //672
    }

    /// Changing values at 'special' addresses can take many more cycles than normal
    pub const fn is_special_memory(address: u16) -> bool {
        matches!(
            address,
            CODE_BANK_1_ID
                | RAM_BANK_1_ID
            |CODE_BANK_2_ID
                | RAM_BANK_2_ID
                | ATLAS1_BANK_ID
                | ATLAS2_BANK_ID
                | ATLAS3_BANK_ID
                | ATLAS4_BANK_ID
                | TIMER_CONTROL
                | SAVE_CONTROL
                | SAVE_BANK_ID
                | SOUND
        )
    }
}

pub mod save_flags {
    pub const AUTO_SAVE: u8 = 4;
    pub const WRITE: u8 = 0;
}

pub mod interrupt_flags {
    pub const IRQ_SCREEN_DRAW: u8 = 1;
    pub const IRQ_TIMER: u8 = 2;
    pub const IRQ_DATETIME: u8 = 4;
    pub const IRQ_CONTROLLER: u8 = 8;
    pub const IRQ_INPUT: u8 = 16;
    pub const IRQ_LINE_DRAW: u8 = 32;
}

#[cfg(test)]
mod test {
    use crate::mem::sizes::{RESERVED, TOTAL};
    use crate::mem::{address, sizes};

    #[test]
    #[allow(clippy::assertions_on_constants)] //these are used as safe guards against changes
    fn test_values() {
        assert_eq!(TOTAL, 65536);
        //the system needs at least 6 bytes available for internal use
        //currently this is just VM::execute_op()
        assert!(RESERVED > 6);
    }

    #[test]
    fn test_sizes_address() {
        assert_eq!(address::CODE, 0);
        assert_eq!(address::CODE_BANK_1, address::CODE + sizes::MAIN_CODE);
        assert_eq!(
            address::CODE_BANK_2,
            address::CODE_BANK_1 + sizes::CODE_BANK
        );
        assert_eq!(address::RAM, address::CODE_BANK_2 + sizes::CODE_BANK);
        assert_eq!(address::RAM_BANK_1, address::RAM + sizes::MAIN_RAM);
        assert_eq!(address::RAM_BANK_2, address::RAM_BANK_1 + sizes::RAM_BANK);
        assert_eq!(address::INPUT, address::RAM_BANK_2 + sizes::RAM_BANK);
        assert_eq!(address::SOUND, address::INPUT + sizes::INPUT);
        assert_eq!(address::SAVE_BANK_ID, address::SOUND + sizes::SOUND);
        assert_eq!(
            address::SAVE_BANK,
            address::SAVE_BANK_ID + sizes::SAVE_BANK_ID
        );
        assert_eq!(address::ATLAS1, address::SAVE_BANK + sizes::SAVE_BANK);
        assert_eq!(address::ATLAS2, address::ATLAS1 + sizes::ATLAS);
        assert_eq!(address::ATLAS3, address::ATLAS2 + sizes::ATLAS);
        assert_eq!(address::ATLAS4, address::ATLAS3 + sizes::ATLAS);
        assert_eq!(address::PALETTES, address::ATLAS4 + sizes::ATLAS);
        assert_eq!(
            address::SPRITE_TABLE,
            address::PALETTES + sizes::PALETTES_TOTAL
        );
        assert_eq!(
            address::LAYER_HEADERS,
            address::SPRITE_TABLE + sizes::SPRITE_TABLE
        );
        assert_eq!(
            address::LAYERS,
            address::LAYER_HEADERS + sizes::LAYERS_HEADER
        );
        assert_eq!(
            address::CODE_BANK_1_ID,
            address::LAYERS + sizes::LAYERS_CONTENT
        );
        assert_eq!(
            address::RAM_BANK_1_ID,
            address::CODE_BANK_1_ID + sizes::CODE_BANK_ID
        );
        assert_eq!(
            address::ATLAS1_BANK_ID,
            address::RAM_BANK_1_ID + sizes::RAM_BANK_ID
        );
        assert_eq!(
            address::ATLAS2_BANK_ID,
            address::ATLAS1_BANK_ID + sizes::ATLAS_BANK_ID
        );
        assert_eq!(
            address::ATLAS3_BANK_ID,
            address::ATLAS2_BANK_ID + sizes::ATLAS_BANK_ID
        );
        assert_eq!(
            address::ATLAS4_BANK_ID,
            address::ATLAS3_BANK_ID + sizes::ATLAS_BANK_ID
        );
        assert_eq!(address::SP, address::ATLAS4_BANK_ID + sizes::ATLAS_BANK_ID);
        assert_eq!(address::FP, address::SP + sizes::SP);
        assert_eq!(address::TIMER_CONTROL, address::FP + sizes::FP);
        assert_eq!(
            address::TIMER_VALUE1,
            address::TIMER_CONTROL + sizes::TIMER_CONTROL
        );
        assert_eq!(
            address::TIMER_VALUE2,
            address::TIMER_VALUE1 + sizes::TIMER_VALUE
        );
        assert_eq!(
            address::TIMER_VALUE3,
            address::TIMER_VALUE2 + sizes::TIMER_VALUE
        );
        assert_eq!(
            address::TIMER_VALUE4,
            address::TIMER_VALUE3 + sizes::TIMER_VALUE
        );
        assert_eq!(
            address::IRQ_RET_ADDR,
            address::TIMER_VALUE4 + sizes::TIMER_VALUE
        );
        assert_eq!(
            address::IRQ_REG_DUMP,
            address::IRQ_RET_ADDR + sizes::IRQ_RET_ADDR
        );
        assert_eq!(
            address::CONTROLLER_PALETTE,
            address::IRQ_REG_DUMP + sizes::IRQ_REG_DUMP
        );
        assert_eq!(
            address::CONTROLLER_TABLE,
            address::CONTROLLER_PALETTE + sizes::CONTROLLER_PALETTE
        );
        assert_eq!(
            address::IRQ_CONTROL,
            address::CONTROLLER_TABLE + sizes::CONTROLLER_TABLE
        );
        assert_eq!(
            address::SAVE_CONTROL,
            address::IRQ_CONTROL + sizes::IRQ_CONTROL
        );
        assert_eq!(
            address::DATETIME,
            address::SAVE_CONTROL + sizes::SAVE_CONTROL
        );
        assert_eq!(address::RAND, address::DATETIME + sizes::DATETIME);
        assert_eq!(address::WAVE_TABLE, address::RAND + sizes::RAND);
        assert_eq!(
            address::CODE_BANK_2_ID,
            address::WAVE_TABLE + sizes::WAVE_TABLE
        );
        assert_eq!(
            address::RAM_BANK_2_ID,
            address::CODE_BANK_2_ID + sizes::CODE_BANK_ID
        );
        assert_eq!(address::VLINE, address::RAM_BANK_2_ID + sizes::RAM_BANK_ID);
        assert_eq!(address::RESERVED, address::VLINE + sizes::VLINE);
        assert_eq!(address::STACK, address::RESERVED + RESERVED);
        assert_eq!(65536, address::STACK as usize + sizes::STACK as usize);
        assert_eq!(address::MAX, 0xFFFF);
    }
}
